tests/ops/op_math/test_adjoint_transform.py

```diff
@@ -0,0 +1,381 @@
+
+    def test_adjoint_on_function(self):
+        """Test adjoint transform on a function"""
+
+        def func(x, y, z):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=0)
+            qml.RZ(z, wires=0)
+
+        x = 1.23
+        y = 2.34
+        z = 3.45
+        with qml.tape.QuantumTape() as tape:
+            out = adjoint(func)(x, y, z)
+
+        assert out == tape.circuit
+
+        for op in tape:
+            assert isinstance(op, Adjoint)
+
+        # check order reversed
+        assert tape[0].base.__class__ is qml.RZ
+        assert tape[1].base.__class__ is qml.RY
+        assert tape[2].base.__class__ is qml.RX
+
+        # check parameters assigned correctly
+        assert tape[0].data == [z]
+        assert tape[1].data == [y]
+        assert tape[2].data == [x]
+
+    def test_nested_adjoint(self):
+        """Test the adjoint transform on an adjoint transform."""
+        x = 4.321
+        with qml.tape.QuantumTape() as tape:
+            out = adjoint(adjoint(qml.RX))(x, wires="b")
+
+        assert out is tape[0]
+        assert isinstance(out, Adjoint)
+        assert isinstance(out.base, Adjoint)
+        assert out.base.base.__class__ is qml.RX
+        assert out.data == [x]
+        assert out.wires == qml.wires.Wires("b")
+
+
+class TestNonLazyExecution:
+    """Test the lazy=False keyword."""
+
+    def test_single_decomposeable_op(self):
+        """Test lazy=False for a single op that gets decomposed."""
+
+        x = 1.23
+        with qml.tape.QuantumTape() as tape:
+            base = qml.RX(x, wires="b")
+            out = adjoint(base, lazy=False)
+
+        assert len(tape) == 1
+        assert out is tape[0]
+        assert tape._queue[base] == {"owner": out}
+        assert tape._queue[out] == {"owns": base}
+
+        assert isinstance(out, qml.RX)
+        assert out.data == [-1.23]
+
+    def test_single_nondecomposable_op(self):
+        """Test lazy=false for a single op that can't be decomposed."""
+        with qml.tape.QuantumTape() as tape:
+            base = qml.S(0)
+            out = adjoint(base, lazy=False)
+
+        assert len(tape) == 1
+        assert out is tape[0]
+        assert len(tape._queue) == 2
+        assert tape._queue[base] == {"owner": out}
+        assert tape._queue[out] == {"owns": base}
+
+        assert isinstance(out, Adjoint)
+        assert isinstance(out.base, qml.S)
+
+    def test_single_decomposable_op_function(self):
+        """Test lazy=False for a single op callable that gets decomposed."""
+        x = 1.23
+        with qml.tape.QuantumTape() as tape:
+            out = adjoint(qml.RX, lazy=False)(x, wires="b")
+
+        assert out is tape[0]
+        assert not isinstance(out, Adjoint)
+        assert isinstance(out, qml.RX)
+        assert out.data == [-x]
+
+    def test_single_nondecomposable_op_function(self):
+        """Test lazy=False for a single op that can't be decomposed."""
```

Comment:
Copy and paste error for docstring?

Reply:
Think that's the right docstring. Is something about it unclear?

Reply:
It's the same docstring to 171...should it be "op callable"?

Reply:
```suggestion
        """Test lazy=False for a single op function that can't be decomposed."""
```
Ah thanks.